tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml

```
@@ -0,0 +1,41 @@
+# It's used for compilation with cache in Mac Host.
+
+parameters:
+- name: TODAY
+  type: string
+
+- name: BuildStep
+  type: stepList
+
+- name: AdditionalKey
+  type: string
+
+- name: CACHE_DIR
+  type: string
+
+steps:
+  - script: |
+      brew install ccache
+      echo "##vso[task.prependpath]/usr/local/opt/ccache/libexec"
+      mkdir -p "${{ parameters.CACHE_DIR }}"
+    displayName: Install ccache and update PATH to use linked versions of gcc, cc, etc
+
+  - task: Cache@2
+    inputs:
+      key:  ' "${{parameters.TODAY}}" | ${{parameters.AdditionalKey}} | "$(Build.SourceVersion)" '
+      path: ${{ parameters.CACHE_DIR }}
+      restoreKeys: |
+        "${{parameters.TODAY}}" | ${{parameters.AdditionalKey}}
+    displayName: ccache task
+
+  - ${{ parameters.BuildStep }}
+
+  - script: |
+      set -ex
+      ccache -s
+      ccache -z
+      ls -l $(Pipeline.Workspace)/ccache
+      du -sh $(Pipeline.Workspace)/ccache
+    displayName: Show Cache stats and Clear stats.
+    env:
+      CCACHE_DIR: ${{parameters.CACHE_DIR}}
```

nit: extra indentation
```suggestion
- script: |
    brew install ccache
    echo "##vso[task.prependpath]/usr/local/opt/ccache/libexec"
    mkdir -p "${{ parameters.CACHE_DIR }}"
  displayName: Install ccache and update PATH to use linked versions of gcc, cc, etc

- task: Cache@2
  inputs:
    key:  ' "${{parameters.TODAY}}" | ${{parameters.AdditionalKey}} | "$(Build.SourceVersion)" '
    path: ${{ parameters.CACHE_DIR }}
    restoreKeys: |
      "${{parameters.TODAY}}" | ${{parameters.AdditionalKey}}
  displayName: ccache task

- ${{ parameters.BuildStep }}

- script: |
    set -ex
    ccache -s
    ccache -z
    ls -l $(Pipeline.Workspace)/ccache
    du -sh $(Pipeline.Workspace)/ccache
  displayName: Show Cache stats and Clear stats.
  env:
    CCACHE_DIR: ${{parameters.CACHE_DIR}}
```